### ISSUE TITLE
[clangd] textDocument/definition stops working after 125 times

### ISSUE THREAD

**Issue by bobajeff** [autozimu/LanguageClient-neovim#101]
After #94 I reported the bug to llvm https://bugs.llvm.org/show_bug.cgi?id=34201 and went on to test if I could reproduce the bug in vscode but I couldn't. So I went on to try and find out some more about the bug and in doing so I've discovered a few things:

- First was that just running `textDocument/completion` by itself was enough to stop `textDocument/definition` from working. So making 2 requests at the same time wasn't causing it.

- Then I've found out that sometimes `textDocument/definition` would still work for a while after running `textDocument/completion`; varying depending on the symbol I ran `textDocument/completion` on. 

- Finally I've found out that if I just run `textDocument/definition` a bunch of times it'll stop working on it's own. _(125 times seems to be the breaking point on my system)_

So I imagine that the real bug it in some buffer in either LanguageClient or neovim which running `textDocument/completion` will fill up quickly while running `textDocument/definition` by itself will fill up slowly but surely.

### Steps to Reproduce:

- open `rplugin/python3/tests/sample-cpp/main.cpp`
- run `LanguageClient_textDocument_definition()` on '`Foo`' at **line 7 column 3** for 125 times _(I created a keymap and also a function to help with this)_.
- after 125 times it'll stop working.

### System info:
- neovim version:
```
NVIM v0.2.0
Build type: Release
Compilation: /usr/bin/cc -march=i686 -mtune=generic -O2 -pipe -fstack-protector-strong -Wconversion -DNVIM_MSGPACK_HAS_FLOAT32 -O2 -DNDEBUG -DDISABLE_LOG -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wvla -fstack-protector-strong -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -I/build/neovim/src/build/config -I/build/neovim/src/neovim-0.2.0/src -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/build/neovim/src/build/src/nvim/auto -I/build/neovim/src/build/include
Compiled by builduser

Optional features included (+) or not (-): +acl   +iconv    +jemalloc +tui      
For differences from Vim, see :help vim-differences

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"
```
- Plugin version
71babc65ee8fb8f17b340e70be2f335777a8f5c0

- Minimal Vimrc
```
call plug#begin('~/.local/share/nvim/plugged')
Plug 'autozimu/LanguageClient-neovim', { 'do': ':UpdateRemotePlugins' }
Plug 'Shougo/deoplete.nvim', { 'do': ':UpdateRemotePlugins' }
call plug#end()

set nu
set title titlestring=%<%F

set hidden
let g:LanguageClient_serverCommands = { 'cpp': ['/home/zarniwhoop/.bin/bin/clangd', '-run-synchronously'] }

let g:deoplete#enable_at_startup = 1
let g:deoplete#disable_auto_complete = 1

inoremap <F12> <C-o>:call LanguageClient_textDocument_definition()<CR>
 
inoremap <silent><expr> <C-Space> deoplete#mappings#manual_complete()
inoremap <silent><expr> <Esc> pumvisible() ? "<C-e>" : "<Esc>"
inoremap <silent><expr> <Cr> pumvisible() ? deoplete#mappings#close_popup() : "<C-g>u<Cr>"

inoremap <F6> <Down><Down><Down><Down><Down><Down><Right><Right><C-o>:call LanguageClient_textDocument_definition()<CR>

function TestDefinition()
let i=0
while i<130
call cursor(7,3) | call LanguageClient_textDocument_definition()
let i = i + 1
endwhile
endfunction

nnoremap <silent> <F7> :call TestDefinition()<CR>
```
- `:CheckHealth`

```
health#deoplete#check
========================================================================
## deoplete.nvim
  - SUCCESS: has("nvim") was successful
  - SUCCESS: has("python3") was successful
  - INFO: If you're still having problems, try the following commands:
    $ export NVIM_PYTHON_LOG_FILE=/tmp/log
    $ export NVIM_PYTHON_LOG_LEVEL=DEBUG
    $ nvim
    $ cat /tmp/log_{PID}
    and then create an issue on github

health#nvim#check
========================================================================
## Configuration
  - SUCCESS: no issues found

## Performance
  - SUCCESS: Build type: Release

## Remote Plugins
  - SUCCESS: Up to date

## terminal
  - INFO: key_backspace (kbs) terminfo entry: key_backspace=^H
  - INFO: key_dc (kdch1) terminfo entry: key_dc=\E[3~

health#provider#check
========================================================================
## Clipboard (optional)
  - SUCCESS: Clipboard tool found: xsel

## Python 2 provider (optional)
  - WARNING: No Python interpreter was found with the neovim module.  Using the first available for diagnostics.
  - WARNING: provider/pythonx: Could not load Python 2:
    /usr/bin/python2 does not have the "neovim" module. :help |provider-python|
    /usr/bin/python2.7 does not have the "neovim" module. :help |provider-python|
    python2.6 not found in search path or not executable.
    /usr/bin/python does not have the "neovim" module. :help |provider-python|
  - ERROR: Python provider error
    - SUGGESTIONS:
      - provider/pythonx: Could not load Python 2:
          /usr/bin/python2 does not have the "neovim" module. :help |provider-python|
          /usr/bin/python2.7 does not have the "neovim" module. :help |provider-python|
          python2.6 not found in search path or not executable.
          /usr/bin/python does not have the "neovim" module. :help |provider-python|
  - INFO: Executable: Not found

## Python 3 provider (optional)
  - INFO: `g:python3_host_prog` is not set.  Searching for python3 in the environment.
  - INFO: Executable: /usr/bin/python3
  - INFO: Python3 version: 3.6.2
  - INFO: python3-neovim version: 0.1.13
  - SUCCESS: Latest python3-neovim is installed: 0.1.13

## Ruby provider (optional)
  - WARNING: `ruby` and `gem` must be in $PATH.
    - SUGGESTIONS:
      - Install Ruby and verify that `ruby` and `gem` commands work.
```

-   Content of `~/.local/share/nvim/rplugin.vim`:
```
" python3 plugins
call remote#host#RegisterPlugin('python3', '/home/zarniwhoop/.local/share/nvim/plugged/LanguageClient-neovim/rplugin/python3/LanguageClient', [
      \ {'sync': v:true, 'name': 'LanguageClient_alive', 'type': 'function', 'opts': {}},
      \ {'sync': v:false, 'name': 'LanguageClient_completionItem/resolve', 'type': 'function', 'opts': {}},
      \ {'sync': v:false, 'name': 'LanguageClient_completionManager_refresh', 'type': 'function', 'opts': {}},
      \ {'sync': v:false, 'name': 'LanguageClient_exit', 'type': 'function', 'opts': {}},
      \ {'sync': v:false, 'name': 'LanguageClient_FZFSinkTextDocumentDocumentSymbol', 'type': 'function', 'opts': {}},
      \ {'sync': v:false, 'name': 'LanguageClient_FZFSinkTextDocumentReferences', 'type': 'function', 'opts': {}},
      \ {'sync': v:false, 'name': 'LanguageClient_FZFSinkWorkspaceSymbol', 'type': 'function', 'opts': {}},
      \ {'sync': v:false, 'name': 'BufReadPost', 'type': 'autocmd', 'opts': {'pattern': '*'}},
      \ {'sync': v:false, 'name': 'CursorMoved', 'type': 'autocmd', 'opts': {'pattern': '*', 'eval': 'line(''.'')'}},
      \ {'sync': v:false, 'name': 'TextChanged', 'type': 'autocmd', 'opts': {'pattern': '*', 'eval': 'fnamemodify(expand("<afile>"), ":p")'}},
      \ {'sync': v:false, 'name': 'TextChangedI', 'type': 'autocmd', 'opts': {'pattern': '*', 'eval': 'fnamemodify(expand("<afile>"), ":p")'}},
      \ {'sync': v:false, 'name': 'VimEnter', 'type': 'autocmd', 'opts': {'pattern': '*'}},
      \ {'sync': v:false, 'name': 'LanguageClient_initialize', 'type': 'function', 'opts': {}},
      \ {'sync': v:false, 'name': 'LanguageClient_registerServerCommands', 'type': 'function', 'opts': {}},
      \ {'sync': v:false, 'name': 'LanguageClient_setLoggingLevel', 'type': 'function', 'opts': {}},
      \ {'sync': v:false, 'name': 'LanguageClientStart', 'type': 'command', 'opts': {'nargs': '*', 'range': ''}},
      \ {'sync': v:false, 'name': 'LanguageClientStop', 'type': 'command', 'opts': {}},
      \ {'sync': v:false, 'name': 'LanguageClient_textDocument_completion', 'type': 'function', 'opts': {}},
      \ {'sync': v:false, 'name': 'LanguageClient_textDocument_completionOmnifunc', 'type': 'function', 'opts': {}},
      \ {'sync': v:false, 'name': 'LanguageClient_textDocument_definition', 'type': 'function', 'opts': {}},
      \ {'sync': v:false, 'name': 'LanguageClient_textDocument_didClose', 'type': 'function', 'opts': {}},
      \ {'sync': v:false, 'name': 'BufWritePost', 'type': 'autocmd', 'opts': {'pattern': '*'}},
      \ {'sync': v:false, 'name': 'LanguageClient_textDocument_documentSymbol', 'type': 'function', 'opts': {}},
      \ {'sync': v:false, 'name': 'LanguageClient_textDocument_formatting', 'type': 'function', 'opts': {}},
      \ {'sync': v:false, 'name': 'LanguageClient_textDocument_hover', 'type': 'function', 'opts': {}},
      \ {'sync': v:false, 'name': 'LanguageClient_textDocument_rangeFormatting', 'type': 'function', 'opts': {}},
      \ {'sync': v:false, 'name': 'LanguageClient_textDocument_references', 'type': 'function', 'opts': {}},
      \ {'sync': v:false, 'name': 'LanguageClient_textDocument_rename', 'type': 'function', 'opts': {}},
      \ {'sync': v:false, 'name': 'LanguageClient_textDocument_signatureHelp', 'type': 'function', 'opts': {}},
      \ {'sync': v:false, 'name': 'LanguageClient_workspace_symbol', 'type': 'function', 'opts': {}},
     \ ])
call remote#host#RegisterPlugin('python3', '/home/zarniwhoop/.local/share/nvim/plugged/deoplete.nvim/rplugin/python3/deoplete', [
      \ {'sync': v:false, 'name': '_deoplete', 'type': 'function', 'opts': {}},
     \ ])


" ruby plugins


" python plugins
```

- LanguageClient.log

[LanguageClient.log.txt](https://github.com/autozimu/LanguageClient-neovim/files/1231649/LanguageClient.log.txt)

**Comment by autozimu** [comment]
Really impressive finding!

**Comment by autozimu** [comment]
Figured the same issue as https://github.com/autozimu/LanguageClient-neovim/issues/21

A bit update though, even after removing `stderr=PIPE` setting, clangd will still stop working due buffer fullness. Seems like the only viable option here is to create a wrapper around clangd.

**Comment by DeltaEvo** [comment]
You can fix it by redirecting output to null :blush: @bobajeff 
```diff
diff --git a/rplugin/python3/LanguageClient/LanguageClient.py b/rplugin/python3/LanguageClient/LanguageClient.py
index 9b50eab..ad0eaf1 100644
--- a/rplugin/python3/LanguageClient/LanguageClient.py
+++ b/rplugin/python3/LanguageClient/LanguageClient.py
@@ -272,7 +272,7 @@ class LanguageClient:
                 command,
                 stdin=subprocess.PIPE,
                 stdout=subprocess.PIPE,
-                stderr=subprocess.PIPE)
+                stderr=open(os.devnull, 'w'))
         except Exception as ex:
             msg = "Failed to start language server: " + ex.args[1]
             logger.exception(msg)
```
However @autozimu for me this beavior is a bug of this project, we have to manager stderr cleanly. I think we should redirect it to a temporary file

**Comment by autozimu** [comment]
I've tried both approach. None of them could fix the issue. Maybe I got
somewhere wrong. Like always, PR is welcomed.

**Comment by bobajeff** [comment]
Thanks alot @DeltaEvo that fix does work me.

@autozimu I'm supprised it doesn't for you. On mine I can now use both `textDocument/completion` and `textDocument/definition` however many times and it'll still work.

In any case it's really a bug in **LanguageClient-neovim** rather than **Clangd** or **Python Language Server**. The correct approach is to change the way LanguageClient is reading from `stdout` and `stderr` so that it doesn't block based on output volume. It shouldn't be doing that and who knows what other issues it's causing.

I'm not familiar with python myself so I'm not sure what the best solution is but I've found a few posts on stackoverflow that tackle the issue _(assuming you don't use the fix by @DeltaEvo)_:
https://stackoverflow.com/questions/375427/non-blocking-read-on-a-subprocess-pipe-in-python
https://stackoverflow.com/questions/36476841/python-how-to-read-stdout-of-subprocess-in-a-nonblocking-way

I believe there needs to be another **Issue** on this until the deadlock is fixed.

**Comment by autozimu** [comment]
There is no deadlock isssue in this case. When there is a completion
request, clangd wtites much more data to stderr, which results in a much
quicker stderr buffer fill.

**Comment by bobajeff** [comment]
@autozimu According to a comment (reposted serveral times) on stackoverflow:

> (Coming from google?) all PIPEs will deadlock when one of the PIPEs' buffer gets filled up and not read. e.g. stdout deadlock when stderr is filled. Never pass a PIPE you don't intend read – Nasser Al-Wohaibi 

It looks like stderr is being filled up and not read. I'm guessing that's why setting it to devnull fixes it (for me anyway).

**Comment by autozimu** [comment]
I know what the problem is, I know how the fix should be. I've been trying more than a dozen times, but it just doesn't fix the issue on my side.

Obviously, I cannot make a commit when even myself cannot verify the fix.

**Comment by bobajeff** [comment]
I'm guessing you're using a different system than mine. (Windows?)

I propose opening up another issue on this just becuase of how complicated the solution might be. Just reading several posts about this on StackOverflow I've seen many variations and proposed solutions. 

Asyncio seems like a promising higher level and portable solution. A example that might fit our case:
https://stackoverflow.com/a/25960956

**Comment by DeltaEvo** [comment]
@autozimu a very simple fix is to redirect stderr to a file like `/tmp/LanguageServer.log`

**Comment by cormacrelf** [comment]
Hm, now I'm getting an error on exactly line 275. Also windows, healthy :CheckHealth.

```
13:03:29 ERROR    Failed to start language server: The system cannot find the file specified
Traceback (most recent call last):
  File "C:/Users/cormacr/AppData/Local/nvim/plugged/LanguageClient-neovim/rplugin/python3\LanguageClient\LanguageClient.py", line 275, in start
    'wb'))
  File "C:\Python35\lib\subprocess.py", line 947, in __init__
    restore_signals, start_new_session)
  File "C:\Python35\lib\subprocess.py", line 1224, in _execute_child
    startupinfo)
FileNotFoundError: [WinError 2] The system cannot find the file specified
```
